### PR TITLE
FIX: milp: return feasible solutions if available on time out

### LINKED ISSUE #16545: BUG: 1.9.0rc1: `OptimizeResult` not populated when `optimize.milp` runs into `time_limit` even though a feasible solution was found.
### Describe your issue.

When `optimize.milp` terminates due to the time limit being reached, the `OptimizeResult` object is not populated, even though one or more feasible solutions have been found.

Example HiGHS log below shows that a feasible solution has been found (see solution status) when the 1 second time limit is reached.
[HiGHS log.txt](https://github.com/scipy/scipy/files/9039809/HiGHS.log.txt)

but the corresponding `OptimizeResult` object (see "Error message" section) is more or less empty and the only thing that is propagated correctly is the `message`, saying that the time limit was reached.

### Reproducing Code Example

```python
import numpy as np
from random import randint
from scipy.optimize import milp, LinearConstraint, Bounds

A = np.zeros((100, 100))
for c in range(100):
    for v in range(100):
        A[c, v] = randint(0, 5)
b_lb = [-1*np.inf for c in range(100)]
b_ub = [25 for c in range(100)]
constraints = LinearConstraint(A, b_lb, b_ub)
variable_lb = [0 for v in range(100)]
variable_ub = [1 for v in range(100)]
variable_bounds = Bounds(variable_lb, variable_ub)
integrality = [1 for v in range(100)]
c_vector = 100 * [-1]
res = milp(
    c_vector,
    integrality=integrality,
    bounds=variable_bounds,
    constraints=constraints,
    options={"time_limit": 1, "disp": True}
)
```

### Error message

```shell
{'fun': None,
 'message': 'Time limit reached. (HiGHS Status 13: model_status is Time limit '
            "reached; primal_status is b'At upper bound')",
 'mip_dual_bound': None,
 'mip_gap': None,
 'mip_node_count': None,
 'status': 1,
 'success': False,
 'x': None}
```

### SciPy/NumPy/Python version information

1.9.0rc1 1.23.0 sys.version_info(major=3, minor=9, micro=13, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by mckib2**
#### Reference issue

Closes gh-16545

#### What does this implement/fix?

- returns any available and feasible solution when the HiGHS MIP solver times out

### CODE REVIEW

File: scipy/optimize/_highs/cython/src/Highs.pxd [hunk 1]
```diff
@@ -30,7 +30,6 @@ cdef extern from "Highs.h":
         # split up for cython below
         #const HighsModelStatus& getModelStatus(const bool scaled_model = False) const
         const HighsModelStatus & getModelStatus() const
-        const HighsModelStatus & getModelStatus(const bool scaled_model) const
```
Comment (mckib2) [inline scipy/optimize/_highs/cython/src/Highs.pxd:33]: This function was deprecated (noticed a deprecation notice in the HiGHS stdout)

File: scipy/optimize/tests/test_milp.py [hunk 2]
```diff
@@ -271,3 +271,31 @@ def test_infeasible_prob_16609():
...
+    res = milp(
+        c_vector,
+        integrality=integrality,
+        bounds=variable_bounds,
+        constraints=constraints,
+        options={"time_limit": .01}
+    )
+    msg = "Time limit reached. (HiGHS Status 13:"
+    assert res.message.startswith(msg)
+    assert res["x"] is not None
```
Comment (mdhaber) [inline scipy/optimize/tests/test_milp.py:301]: Could you show feasibility?
And can we parametrize to show that we get a solution when the iteration limit is reached?

File: scipy/optimize/_highs/cython/src/_highs_wrapper.pyx [hunk 3]
```diff
@@ -705,7 +722,6 @@ def _highs_wrapper(
         res = {
             'status': <int> model_status,
             'message': highs.modelStatusToString(model_status).decode(),
-            'unscaled_status': <int> unscaled_model_status,
```
Comment (mdhaber) [inline scipy/optimize/_highs/cython/src/_highs_wrapper.pyx:708]: This is unrelated? Can you remind me what this is for and why it's no longer needed?

Comment (mckib2) [reply scipy/optimize/_highs/cython/src/_highs_wrapper.pyx:708]: This is related to the note above about deprecation of `getModelStatus(const bool scaled_model)`.  The scaled `model_status` used to be compared to the normal `unscaled_model_status` to know if there was any discrepancies between the scaled problem vs original problem, but this is now absent in the main HiGHS repo and all instances of `scaled_model_status` have been removed.  The function was moved to a deprecated functions file and `unscaled_model_status` will always be equal to `model_status`, so it made sense to me to remove it from here as it adds no information

File: scipy/optimize/tests/test_milp.py (hunk 2 above)
Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:301]: The other time limit test does assert that `res['x'] is None` when it has not found a feasible solution.  I can add in a check to ensure feasibility, yes

File: scipy/optimize/tests/test_milp.py [hunk 4]
```diff
@@ -271,3 +271,38 @@ def test_infeasible_prob_16609():
...
+        for v in range(100):
+            A[c, v] = np.random.randint(0, 5)
+    b_lb = np.array([-1 * np.inf for c in range(100)])
+    b_ub = np.array([25 for c in range(100)])
+    constraints = LinearConstraint(A, b_lb, b_ub)
+    variable_lb = [0 for v in range(100)]
+    variable_ub = [1 for v in range(100)]
+    variable_bounds = Bounds(variable_lb, variable_ub)
+    integrality = [1 for v in range(100)]
+    c_vector = 100 * [-1]
```
Comment (mdhaber) [inline scipy/optimize/tests/test_milp.py:292]:
```suggestion
    rng = np.random.default_rng(5123833489170494244)
    A = rng.integers(0, 5, size=(100, 100))
    b_lb = np.full(100, fill_value=-np.inf)
    b_ub = np.full(100, fill_value=25)
    constraints = LinearConstraint(A, b_lb, b_ub)
    variable_lb = np.zeros(100)
    variable_ub = np.ones(100)
    variable_bounds = Bounds(variable_lb, variable_ub)
    integrality = np.ones(100)
    c_vector = -np.ones(100)
```
using `np.random.default_rng` instead of `randint` to make Pamphile happy.
Suggesting the rest for concision.

**Review by mdhaber — COMMENTED** [review]
Had a chance to test this locally and it seems perfect. If these refactoring suggestions look good to you, please commit them. Also let me know about whether HiGHS exposes an iteration limit option so we can add that to the test.

File: scipy/optimize/tests/test_milp.py (hunk 2 above)
Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:299]: Looking into iteration limit now

Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:301]: I went ahead and added `mip_max_nodes` as a first-class option (hopefully didn't jump the gun on this one) and added the suggest parameterization to the test

Comment (mdhaber) [reply scipy/optimize/tests/test_milp.py:301]: Is that the thing that corresponds with `HighsModelStatusREACHED_ITERATION_LIMIT`?

Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:301]: > Is that the thing that corresponds with `HighsModelStatusREACHED_ITERATION_LIMIT`? That's the part of the changes I wanted to test.

Yes, `mip_max_nodes` being reached will stop the MIP solver and return `HighsModelStatusREACHED_ITERATION_LIMIT`

Comment (mdhaber) [reply scipy/optimize/tests/test_milp.py:302]: Is `mip_max_nodes` what HiGHS calls the option? If so, I guess it's fine. I just think it will be surprising to a user that if they set `mip_max_nodes` too low, they get an issue about reaching an "iteration" limit.

Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:302]: `mip_max_nodes` is the HiGHS option, but we can alias it to something else if it will be clearer.  From a cursory look around, "iteration limit" almost always applies to the simplex solvers, not the MIP solver (e.g., [Gurobi](https://www.gurobi.com/documentation/9.5/refman/parameters.html) has both an `IterationLimit` and `NodeLimit`), so I think it might be confusing to use "iteration limit"

Comment (mdhaber) [reply scipy/optimize/tests/test_milp.py:305]: OK, I made it `node_limit`.

**Comment by mdhaber** [comment]
Renamed to `node_limit`. It seems that if `mip_max_nodes` is `None`, it behaves the same was as if `mip_max_nodes` is not present in the options structure at all? If so, I think this is all set. Merged main, and if CI passes, I'll go ahead and merge.

**Comment by mdhaber** [comment]
Hmm looks like the tests are really failing on some platforms. One is workers crashing and the other looks like a 60s timeout on the iteration limit test. Maybe reduce the problem size?

**Comment by mckib2** [comment]
> It seems that if mip_max_nodes is None, it behaves the same was as if mip_max_nodes is not present in the options structure at all?

Correct -- if `None` then the HiGHS default values are assumed (in this case, no limit)

> One is workers crashing and the other looks like a 60s timeout on the iteration limit test. Maybe reduce the problem size?

If it's problem size, reducing the `node_limit` may also be an option.  I'll look at the logs

**Comment by mckib2** [comment]
@mdhaber As I started looking into this, things looked more and more familiar until I came across these two old, related issues:

- https://github.com/ERGO-Code/HiGHS/issues/769
- gh-16347

Appears we and Julia maintainers have run into this issue before and our strategy was to ignore those tests for 32-bit builds.  Is that still an acceptable strategy here?

**Comment by mdhaber** [comment]
Yes, that's fine with me.

**Comment by mckib2** [comment]
> Yes, that's fine with me. Is HiGHS planning on fixing it?

Not currently, but I think I found a solution they might take upstream to fix this and gh-16347.  This is apparently a notorious GCC floating point error bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=323#c223.  I reading up now on what the best workaround in this case will be

**Comment by mckib2** [comment]
xref https://github.com/ERGO-Code/HiGHS/pull/944

The upstream patch I've just submitted resolves all the 32-bit issues we've encountered with `milp` (tested on 32-bit xubuntu 18 VM)

**Comment by mdhaber** [comment]
OK, for now let's xfail the test though.

File: scipy/optimize/tests/test_milp.py [hunk 5]
```diff
@@ -273,6 +273,46 @@ def test_infeasible_prob_16609():
     np.testing.assert_equal(res.status, 2)
 
 
+_msg_time = "Time limit reached. (HiGHS Status 13:"
+_msg_iter = "Iteration limit reached. (HiGHS Status 14:"
+
+
+@pytest.mark.skipif(np.intp(0).itemsize < 8, reason="Unhandled 32-bit GCC FP bug")
```
Comment (mdhaber) [inline scipy/optimize/tests/test_milp.py:280]:
```suggestion
@pytest.mark.skipif(np.intp(0).itemsize < 8,
                    reason="Unhandled 32-bit GCC FP bug")
```

**Review by mdhaber — COMMENTED** [review]
Besides the PEP8 issue, the timeout test failure seems real, as though the options are not being respected in that platform?
Besides fixing that, I think we'll want to adjust the tutorial with the binary string issue because that will start failing if this fixes the problem.

File: scipy/optimize/tests/test_milp.py (hunk 5 above)
Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:280]: I'll dig deeper into the `prerelease_deps_coverage_64bit_blas` failure this evening.  A cursory reading of test logs indicate the time out test did indeed report a time out status (though it's not clear if it did after 1 second as desired) but failed to find a solution and the node limited test had pytest time it out (`Failed: Timeout >60.0s`)

Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:280]: When I try to follow the azure-travis-template.yml steps for this job locally in a VM, I fail to reproduce the errors.  I get another unrelated test failure (`scipy/sparse/linalg/_dsolve/tests/test_linsolve.py::TestFactorized::test_bug_8278`), but these milp tests run fine.

Any ideas on what to try next to reproduce/resolve this?

**Comment by mckib2** [comment]
After writing [this comment](https://github.com/scipy/scipy/pull/16897#issuecomment-1229232147), I wonder if the current test failure is due to a thread-safety issue and we're seeing a failure due to those two tests run in parallel.  I'll hopefully be able to look at this more over the next couple of days

**Comment by mdhaber** [comment]
Hope the thought pans out. Unfortunately, I don't have any ideas. Can you reproduce it on one of your platforms?

File: scipy/optimize/tests/test_milp.py [hunk 6]
```diff
@@ -273,6 +273,47 @@ def test_infeasible_prob_16609():
     np.testing.assert_equal(res.status, 2)
 
 
+_msg_time = "Time limit reached. (HiGHS Status 13:"
+_msg_iter = "Iteration limit reached. (HiGHS Status 14:"
+
+
+@pytest.mark.skipif(np.intp(0).itemsize < 8,
+                    reason="Unhandled 32-bit GCC FP bug")
+@pytest.mark.slow
```
Comment (mdhaber) [inline scipy/optimize/tests/test_milp.py:282]:
```suggestion
```

**Comment by mdhaber** [comment]
@mckib2 I removed the `slow` mark to see if this is failing on any other platform.

**Comment by mdhaber** [comment]
This is currently not working on some platforms, so that needs to be solved. 
That said, we could merge this (skipping the test on failing platforms) and leave the underlying issue open.

**Comment by laenNoCode** [comment]
can you try forwarding the changes that were done on the main branch ? I tried yesterday and the failing unit tests now pass, so maybe it could solve the build situation

**Comment by mckib2** [comment]
> can you try forwarding the changes that were done on the main branch ? I tried yesterday and the failing unit tests now pass, so maybe it could solve the build situation

@laenNoCode Do you mean pull in the latest SciPy main or HiGHS main branch?  I just pulled in the latest SciPy and kicked off the CI

**Comment by laenNoCode** [comment]
> > can you try forwarding the changes that were done on the main branch ? I tried yesterday and the failing unit tests now pass, so maybe it could solve the build situation
> 
> @laenNoCode Do you mean pull in the latest SciPy main or HiGHS main branch? I just pulled in the latest SciPy and kicked off the CI

I had other broken tests at the time, so getting the latest SciPy version fixed all tests on my machine. The problem on azure build may be that the HiGHS haven't got the time to find at least one feasible solution, so my guess is that info.objective_function_value == HIGHS_CONST_INF is true
or (model_status in {
        HighsModelStatusREACHED_TIME_LIMIT,
        HighsModelStatusREACHED_ITERATION_LIMIT,
    } and (info.objective_function_value == HIGHS_CONST_INF)).
I'm not familiar with azure, so i don't know how to reproduce the problem and test eventual fixes. Maybe can you link me a tutorial to do so ?

**Comment by mdhaber** [comment]
> The problem on azure build may be that the HiGHS haven't got the time to find at least one feasible solution

I'm not sure that's suggested by the failure message. The test exceeds the pytest time limit(>60s), so HiGHS seems to be ignoring the time limit option entirely. 

Click on "Details". It takes a few more clicks to get to the results, but I'm not at a computer right now to send screenshots.

> I'm not familiar with azure, so i don't know how to reproduce the problem and test eventual fixes

I think everyone would like that, but I'm not aware of a way.

**Comment by mdhaber** [comment]
image

image

image

image

```
scipy/optimize/_milp.py:362: in milp
    highs_res = _highs_wrapper(c, indptr, indices, data, b_l, b_u,
E   Failed: Timeout >60.0s
        args_iv    = (array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1., -1....7530, 7607, 7683, 7759, 7836,
       7913, 7999], dtype=int32), array([ 0,  1,  2, ..., 97, 98, 99], dtype=int32), ...)
        b_l        = array([-inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf,
       -inf, -inf, -inf, -inf, -inf, -inf, -i...f, -inf, -inf, -inf, -inf, -inf,
       -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf, -inf,
       -inf])
```

**Comment by laenNoCode** [comment]
image
https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=20602&view=logs&j=7902a2c1-7b88-5b2d-6408-84b88759c642&t=5471a576-12f4-5533-4de4-506800a17112&l=937
This is the output i based my reasonning on (build log #L937)
on line 776

**Comment by laenNoCode** [comment]
I was able to make the tests pass on azure by allowing more time and by having a lower node_limit. The problem appears to be that azure is way too slow, hence the timeout on node_limit and the lack of answers on time_limit :
image

image

File: scipy/optimize/tests/test_milp.py [hunk 7]
```diff
@@ -273,6 +273,47 @@ def test_infeasible_prob_16609():
...
+@pytest.mark.parametrize(["options", "msg"], [({"time_limit": 1}, _msg_time),
+                                              ({"node_limit": 10}, _msg_iter)])
```
Comment (mdhaber) [inline scipy/optimize/tests/test_milp.py:284]:
```suggestion
@pytest.mark.parametrize(["options", "msg"], [({"time_limit": 10}, _msg_time),
                                              ({"node_limit": 1}, _msg_iter)])
```

**Comment by mdhaber** [comment]
I see. Good catch @laenNoCode. I didn't notice that both tests were failing. 
(How did you test on Azure, BTW?)
I updated this to see how that goes.
@mckib2, if this passes, are you happy with this?

File: scipy/optimize/tests/test_milp.py (hunk 6 above)
Comment (laenNoCode) [inline scipy/optimize/tests/test_milp.py:282]:
```suggestion
@pytest.mark.slow
@pytest.mark.timeout(360) 
```
i added a greater timeout to allow azure to find at least one feasible solution

**Comment by laenNoCode** [comment]
> I see. Good catch @laenNoCode. I didn't notice that both tests were failing. (How did you test on Azure, BTW?) I updated this to see how that goes. @mckib2, if this passes, are you happy with this?

I so first i cloned the repository. Then i created an azure account. After that I created a project on my azure account referencing my repository. I then filled the form for getting free tier CI https://aka.ms/azpipelines-parallelism-request after a day i could start private azure build for the tests

File: scipy/optimize/tests/test_milp.py (hunk 6 above)
Comment (laenNoCode) [reply scipy/optimize/tests/test_milp.py:282]: The test checks if limiting via iteration also works properly but this take too much time. Maybe we should have a better test to check that (one that is faster) but this should fix azure tests

**Comment by mdhaber** [comment]
I see that this is very important to you : ) We'll try to make sure this gets fixed for 1.9.2!

**Comment by laenNoCode** [comment]
> > @mckib2, if this passes, are you happy with this?
> 
> Yes, but looks like some CI machines are getting iteration limit reached instead of timeout. This could be inherent flake to the test, might need to lower timeout or raise iteration limit (but this may undue some of the work needed to get other tests passing). Another adjustment could be to allow both timeout and iteration limit reached as acceptable results.
> 
> The other failure is a refguide fix because of the string encoding fix, should be simple.
> 
> I will look more this evening if someone doesn't get there before me.



> > @mckib2, if this passes, are you happy with this?
> 
> Yes, but looks like some CI machines are getting iteration limit reached instead of timeout. This could be inherent flake to the test, might need to lower timeout or raise iteration limit (but this may undue some of the work needed to get other tests passing). Another adjustment could be to allow both timeout and iteration limit reached as acceptable results.
> 
> The other failure is a refguide fix because of the string encoding fix, should be simple.
> 
> I will look more this evening if someone doesn't get there before me.

if i read the fix well, it also add support and test for early return on iteration limit reach : 
image

image
It appears that the iteration limit test is too slow, hence my proposition to either add more time to allow highs solver to reach iteration limit (https://github.com/scipy/scipy/pull/16814#discussion_r984932964) or to change the iteration limit reach test to a simpler problem to solve.
For the refguide fix i don't know where it is.

**Comment by laenNoCode** [comment]
> I see that this is very important to you : ) We'll try to make sure this gets fixed for 1.9.2!
indeed, i use this on my day to day work. To rebuild scipy on every commit makes the CI/CD process quite long.
Thank you very much !

**Comment by j-bowhay** [comment]
> > > @mckib2, if this passes, are you happy with this?
> > 
> > 
> > Yes, but looks like some CI machines are getting iteration limit reached instead of timeout. This could be inherent flake to the test, might need to lower timeout or raise iteration limit (but this may undue some of the work needed to get other tests passing). Another adjustment could be to allow both timeout and iteration limit reached as acceptable results.
> > The other failure is a refguide fix because of the string encoding fix, should be simple.
> > I will look more this evening if someone doesn't get there before me.
> 
> if i read the fix well, it also add support and test for early return on iteration limit reach : image
> 
> image It appears that the iteration limit test is too slow, hence my proposition to either add more time to allow highs solver to reach iteration limit ([#16814 (comment)](https://github.com/scipy/scipy/pull/16814#discussion_r984932964)) or to change the iteration limit reach test to a simpler problem to solve. For the refguide fix i don't know where it is.

Looks like the tutorial needs a minor update:
```
/home/vsts/work/1/s/tools/../doc/source/tutorial/optimize.rst
-------------------------------------------------------------

File "doc/source/tutorial/optimize.rst", line 1537, in /home/vsts/work/1/s/tools/../doc/source/tutorial/optimize.rst
Failed example:
    print(result.message)
Expected:
    The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is b'At lower/fixed bound')
Got:
    The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is At lower/fixed bound)

```

File: scipy/optimize/tests/test_milp.py (hunk 6 above)
Comment (mckib2) [reply scipy/optimize/tests/test_milp.py:282]: If we can think up another problem that has the property of consistently showing iteration limit reached/timeout reached then we can swap it in (or I suppose we could split it into two tests if we needed different problems).  Happy to raise the timeout to see if that fixes it -- this is a "slow" test after all

File: doc/source/tutorial/optimize.rst [hunk 8]
```diff
@@ -1535,7 +1535,7 @@ Finally, we can solve the transformed problem using :func:`linprog`.
     >>> bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]
     >>> result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)
     >>> print(result.message)
-    The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is b'At lower/fixed bound')
+    The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is At lower/fixed bound)
```
Comment (laenNoCode) [inline doc/source/tutorial/optimize.rst:1538]:
```suggestion
    The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is b'At lower/fixed bound')
```

Comment (j-bowhay) [reply doc/source/tutorial/optimize.rst:1538]: Not sure this change is correct the refguide is now passing

Comment (mckib2) [reply doc/source/tutorial/optimize.rst:1538]: This was corrected in 89c2d019cd818dbdf77cfbbda0661fde86fd326f

**Comment by mdhaber** [comment]
OK, the timeout fixed this. @mckib2 are you happy with this? If so, let's get this in so it can be part of 1.9.2.

**Comment by mckib2** [comment]
Thumbs up from me
